File: TUnit.TestProject/Bugs/1924/DataClass.cs
using TUnit.Core.Interfaces;

namespace TUnit.TestProject.Bugs._1924;

public class DataClass : IAsyncInitializer, IAsyncDisposable
{
    public bool Disposed { get; private set; }

    public Task InitializeAsync()
    {
        return Task.CompletedTask;
    }

    public ValueTask DisposeAsync()
    {
        Disposed = true;
        return default;
    }

    public ValueTask DoSomething()
    {
        if (Disposed)
        {
            throw new ObjectDisposedException(nameof(DataClass));
        }
        
        return default;
    }
}

File: tests/TUnit.Templates.Tests/Snapshots/InstantiationTestWithFSharpDotCover.TUnit.FSharp.--enable-dotcover.verified/TUnit.FSharp/Data/DataClass.fs
namespace TUnit.FSharp

open System
open System.Threading.Tasks
open TUnit.Core.Interfaces

type DataClass() =
    interface IAsyncInitializer with
        member _.InitializeAsync() : Task =
            Console.Out.WriteLineAsync("Classes can be injected into tests, and they can perform some initialisation logic such as starting an in-memory server or a test container.")

    interface IAsyncDisposable with
        member _.DisposeAsync() : ValueTask =
            ValueTask(Console.Out.WriteLineAsync("And when the class is finished with, we can clean up any resources."))

